lib/burnside/intrinsics.cc

```diff
@@ -448,4 +460,61 @@ mlir::Value *IntrinsicLibrary::Implementation::genConjg(
   return cplxHandler.insert<ComplexHandler::Part::Imag>(cplx, negImag);
 }
 
+static mlir::FuncOp getMergeFunc(mlir::Type type, mlir::ModuleOp module) {
+  llvm::SmallVector<mlir::Type, 3> argTypes{
+      type, type, fir::LogicalType::get(module.getContext(), 4)};
+  auto mergeType{
+      mlir::FunctionType::get(argTypes, {type}, module.getContext())};
+  auto mergeName{getIntrinsicWrapperName("merge", mergeType)};
+  return createFunction(module, mergeName, mergeType);
+}
+
+template<Extremum extremum>
+static mlir::Value *createCompare(mlir::Location loc, mlir::OpBuilder &builder,
+    mlir::Value *left, mlir::Value *right) {
+  static constexpr auto integerPredicate{extremum == Extremum::Max
+          ? mlir::CmpIPredicate::SGT
+          : mlir::CmpIPredicate::SLT};
+  static constexpr auto realPredicate{extremum == Extremum::Max
+          ? fir::CmpFPredicate::UGT
+          : fir::CmpFPredicate::ULT};
+  auto type{left->getType()};
+  mlir::Value *result{nullptr};
+  if (type.isa<mlir::FloatType>() || type.isa<fir::RealType>()) {
+    // TODO: The type we get from CmpfOp is a bit weird (it is the operands
+    // types, and not a boolean).
+    result = builder.create<fir::CmpfOp>(loc, realPredicate, left, right);
+  } else if (type.isa<mlir::IntegerType>()) {
+    result = builder.create<mlir::CmpIOp>(loc, integerPredicate, left, right);
+  } else if (type.isa<fir::CharacterType>()) {
+    // TODO: ! character min and max is tricky because the result
+    // length is the length of the longest argument!
+    // So we may need a temp.
+  }
+  assert(result);
+  // TODO which logical type should comparisons return ?
+  // This is also currently not observed in convert-expr
```

Comment:
I'm thinking that we want to use `i8` everywhere except on the effecting periphery (load, store, arguments, ...).  The reason there is that compares are often used with conditional branches, which take an `i8`.  I don't think it changes the result of logical primitives.  `a & b` is the same if `a` and `b` are `i8` or `logical<1>`, I suspect.

Reply:
I do not think either. I am not familiar enough with Fortran logical types. I only know their storage size is influenced by their kind (default logical types and default integer must both occupy a single numeric storage unit) (19.5.3.2). This requirement seems compatible with your proposal to only observe the Fortran logical types around storage operations.